File: LifeTrackerBackend/Constants/GameConstants.cs
namespace LifeTracker.Constants;

public static class GameConstants
{
    public const int BaseXp = 100;
    public const double XpExponent = 1.8;
    public const double ScaleFactor = 50.0;
    
    public const int BaseHp = 50;
    public const int HpPerLevel = 5;
    public const int MaxLevel = 999;
    
    public const double DeathHpResetPercent = 0.25;
    public const double DeathXpPenaltyPercent = 0.10;    
    public const double DeathGoldPenaltyPercent = 0.20;  
    public const double DeathStreakPenaltyPercent = 0.50; 
    public const int RecoveryDebuffHours = 4;            
    public const double RecoveryDebuffMultiplier = 0.75; 
    
    public const double StreakMultiplierCoeff = 0.15;
    public const int MaxFreezeCharges = 3;
    public const int StreakTierDays = 30;
    
    public const int DailyTaskCap = 50;
    
    public static double GetDifficultyMultiplier(TaskDifficulty difficulty) => difficulty switch
    {
        TaskDifficulty.Easy => 1.0,
        TaskDifficulty.Medium => 1.5,
        TaskDifficulty.Hard => 2.5,
        TaskDifficulty.Epic => 4.0,
        _ => 1.0
    };
    
    public static (int xp, int gold) GetHabitReward(TaskDifficulty difficulty) => difficulty switch
    {
        TaskDifficulty.Easy => (10, 5),
        TaskDifficulty.Medium => (25, 12),
        TaskDifficulty.Hard => (50, 25),
        TaskDifficulty.Epic => (100, 50),
        _ => (10, 5)
    };
    
    public static (int xp, int gold) GetOneTimeReward(TaskDifficulty difficulty) => difficulty switch
    {
        TaskDifficulty.Easy => (15, 8),
        TaskDifficulty.Medium => (35, 18),
        TaskDifficulty.Hard => (70, 35),
        TaskDifficulty.Epic => (150, 75),
        _ => (15, 8)
    };
    
    public static (int hpLoss, int goldLoss) GetHabitPenalty(TaskDifficulty difficulty) => difficulty switch
    {
        TaskDifficulty.Easy => (5, 0),
        TaskDifficulty.Medium => (10, 5),
        TaskDifficulty.Hard => (20, 15),
        TaskDifficulty.Epic => (35, 30),
        _ => (5, 0)
    };
    
    public static (int hpLoss, int goldLoss) GetOneTimePenalty(TaskDifficulty difficulty) => difficulty switch
    {
        TaskDifficulty.Easy => (3, 0),
        TaskDifficulty.Medium => (7, 5),
        TaskDifficulty.Hard => (15, 15),
        TaskDifficulty.Epic => (25, 30),
        _ => (3, 0)
    };
    
    public static (int xpPenalty, int goldPenalty, int cooldownHours) GetStreakBreakPenalty(int streakDays) => streakDays switch
    {
        <= 7 => (0, 0, 0),
        <= 30 => (50, 25, 24),
        <= 90 => (150, 75, 48),
        _ => (300, 150, 72)
    };
    
    public static long CalculateXpForLevel(int level) =>
        (long)Math.Floor(BaseXp * Math.Pow(level, XpExponent) * (1 + level / ScaleFactor));
    
    public static int CalculateMaxHp(int level) => BaseHp + (HpPerLevel * level);
    
    public static double CalculateStreakMultiplier(int streakDays) =>
        streakDays <= 0 ? 1.0 : 1.0 + Math.Log2(streakDays + 1) * StreakMultiplierCoeff;
    
    public static double CalculateLevelScaling(int heroLevel) => 1.0 + (heroLevel / 100.0);
}

public enum TaskDifficulty
{
    Easy = 1,
    Medium = 2,
    Hard = 3,
    Epic = 4
}

File: LifeTrackerMobile/app/src/main/java/com/lifetracker/mobile/constants/GameConstants.kt
package com.lifetracker.mobile.constants

import com.lifetracker.mobile.domain.model.TaskDifficulty
import kotlin.math.floor
import kotlin.math.log2
import kotlin.math.pow

object GameConstants {
    const val BASE_XP: Int = 100
    const val XP_EXPONENT: Double = 1.8
    const val SCALE_FACTOR: Double = 50.0

    const val BASE_HP: Int = 50
    const val HP_PER_LEVEL: Int = 5
    const val MAX_LEVEL: Int = 999

    const val DEATH_HP_RESET_PERCENT: Double = 0.25
    const val DEATH_XP_PENALTY_PERCENT: Double = 0.10
    const val DEATH_GOLD_PENALTY_PERCENT: Double = 0.20
    const val DEATH_STREAK_PENALTY_PERCENT: Double = 0.50
    const val RECOVERY_DEBUFF_HOURS: Int = 4
    const val RECOVERY_DEBUFF_MULTIPLIER: Double = 0.75

    const val STREAK_MULTIPLIER_COEFF: Double = 0.15
    const val MAX_FREEZE_CHARGES: Int = 3
    const val STREAK_TIER_DAYS: Int = 30

    const val DAILY_TASK_CAP: Int = 50

    data class Reward(val xp: Int, val gold: Int)
    data class Penalty(val hpLoss: Int, val goldLoss: Int)
    data class StreakBreakPenalty(val hpLoss: Int, val goldLoss: Int, val debuffHours: Int)

    fun getDifficultyMultiplier(difficulty: TaskDifficulty): Double = when (difficulty) {
        TaskDifficulty.Easy   -> 1.0
        TaskDifficulty.Medium -> 1.5
        TaskDifficulty.Hard   -> 2.5
        TaskDifficulty.Epic   -> 4.0
    }

    fun getHabitReward(difficulty: TaskDifficulty): Reward = when (difficulty) {
        TaskDifficulty.Easy   -> Reward(xp = 10, gold = 5)
        TaskDifficulty.Medium -> Reward(xp = 25, gold = 12)
        TaskDifficulty.Hard   -> Reward(xp = 50, gold = 25)
        TaskDifficulty.Epic   -> Reward(xp = 100, gold = 50)
    }

    fun getOneTimeReward(difficulty: TaskDifficulty): Reward = when (difficulty) {
        TaskDifficulty.Easy   -> Reward(xp = 15, gold = 8)
        TaskDifficulty.Medium -> Reward(xp = 35, gold = 18)
        TaskDifficulty.Hard   -> Reward(xp = 70, gold = 35)
        TaskDifficulty.Epic   -> Reward(xp = 150, gold = 75)
    }

    fun getHabitPenalty(difficulty: TaskDifficulty): Penalty = when (difficulty) {
        TaskDifficulty.Easy   -> Penalty(hpLoss = 5, goldLoss = 0)
        TaskDifficulty.Medium -> Penalty(hpLoss = 10, goldLoss = 5)
        TaskDifficulty.Hard   -> Penalty(hpLoss = 20, goldLoss = 15)
        TaskDifficulty.Epic   -> Penalty(hpLoss = 35, goldLoss = 30)
    }

    fun getOneTimePenalty(difficulty: TaskDifficulty): Penalty = when (difficulty) {
        TaskDifficulty.Easy   -> Penalty(hpLoss = 3, goldLoss = 0)
        TaskDifficulty.Medium -> Penalty(hpLoss = 7, goldLoss = 5)
        TaskDifficulty.Hard   -> Penalty(hpLoss = 15, goldLoss = 15)
        TaskDifficulty.Epic   -> Penalty(hpLoss = 25, goldLoss = 30)
    }

    fun getStreakBreakPenalty(streakDays: Int): StreakBreakPenalty = when {
        streakDays <= 7  -> StreakBreakPenalty(hpLoss = 0, goldLoss = 0, debuffHours = 0)
        streakDays <= 30 -> StreakBreakPenalty(hpLoss = 50, goldLoss = 25, debuffHours = 24)
        streakDays <= 90 -> StreakBreakPenalty(hpLoss = 150, goldLoss = 75, debuffHours = 48)
        else             -> StreakBreakPenalty(hpLoss = 300, goldLoss = 150, debuffHours = 72)
    }

    fun calculateXpForLevel(level: Int): Long =
        floor(BASE_XP * level.toDouble().pow(XP_EXPONENT) * (1 + level / SCALE_FACTOR)).toLong()

    fun calculateMaxHp(level: Int): Int = BASE_HP + (HP_PER_LEVEL * level)

    fun calculateStreakMultiplier(streakDays: Int): Double =
        if (streakDays <= 0) 1.0
        else 1.0 + log2((streakDays + 1).toDouble()) * STREAK_MULTIPLIER_COEFF

    fun calculateLevelScaling(heroLevel: Int): Double = 1.0 + (heroLevel / 100.0)
}